cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -191,38 +191,57 @@ VAR_LEN_TYPES(IS_NOT_NULL, isnotnull)
 #undef IS_NOT_NULL
 
 FORCE_INLINE
-const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
-                                    int64 offset64, int64 length, int32* out_len) {
-  if (length <= 0 || input == nullptr || in_len <= 0) {
-    *out_len = 0;
+const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_data_len,
+                                    int64 offset64, int64 length, int32* out_data_len) {
+  if (length <= 0 || input == nullptr || in_data_len <= 0) {
+    *out_data_len = 0;
     return "";
   }
 
+  int32 num_in_chars;  // num of chars in input string
+  int32* char_start =
```

Comment:
Can't this problem be solve in O(1) temporary data?

Reply:
yes, should go over input bytes twice though since offset can be negative.

Reply:
what does it mean for the offset to be negative.

Reply:
negative offsets are counted from string end. For example, substr('arrow', -3, 3) = 'row'

Reply:
going over data twice is cheaper than memory allocation.

Reply:
done. thanks for reviewing